File: backend-go/internal/modules/system/service/login_log_service.go
package service

import (
	"time"

	"drip-admin/backend-go/internal/common"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

func (s *Server) writeLoginLog(c *gin.Context, userID *common.Int64String, username string, realName *string, loginType string, status string, reason string, deviceType string) {
	row := newLoginLogRow(c, userID, username, realName, loginType, status, reason, deviceType)
	if s.db == nil {
		s.logger.Error("login log write failed", zap.String("reason", "database is not configured"), zap.String("username", username), zap.String("status", status))
		return
	}
	if err := s.db.Create(&row).Error; err != nil {
		s.logger.Error("login log write failed", zap.String("username", username), zap.String("status", status), zap.Error(err))
	}
}

func newLoginLogRow(c *gin.Context, userID *common.Int64String, username string, realName *string, loginType string, status string, reason string, deviceType string) SysLoginLog {
	var failureReason *string
	if reason != "" {
		failureReason = &reason
	}
	ip := clientIP(c)
	userAgent := c.GetHeader("User-Agent")
	now := time.Now()
	return SysLoginLog{
		ID:            common.NewID(),
		UserID:        userID,
		Username:      username,
		RealName:      realName,
		LoginType:     loginType,
		Status:        status,
		FailureReason: failureReason,
		IP:            &ip,
		UserAgent:     &userAgent,
		DeviceType:    &deviceType,
		LoginAt:       &now,
	}
}

File: backend-python/app/modules/system/service/login_log_service.py
from datetime import datetime
import logging

from sqlalchemy.ext.asyncio import AsyncSession

from app.common.id import new_id
from app.modules.system.entity import SysLoginLog

logger = logging.getLogger("drip.login")


class LoginLogService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def write(
        self,
        user_id: int | None,
        username: str,
        real_name: str | None,
        login_type: str,
        status: str,
        failure_reason: str | None,
        ip: str,
        user_agent: str,
        device_type: str,
    ) -> None:
        try:
            self.db.add(
                SysLoginLog(
                id=new_id(),
                user_id=user_id,
                username=username,
                real_name=real_name,
                login_type=login_type,
                status=status,
                failure_reason=failure_reason,
                ip=ip,
                user_agent=user_agent,
                device_type=device_type,
                login_at=datetime.now(),
                )
            )
            await self.db.commit()
        except Exception:
            await self.db.rollback()
            logger.exception("login log write failed username=%s status=%s", username, status)
